### PR TITLE
Allow some state tests to be skipped

### PR DESCRIPTION
**Pull request by izaaklauer**
Currently, there's no good way for any other project consuming these tests to selectively skip any of them. Maybe there will be [as of go 1.19](https://github.com/golang/go/issues/41583), but until then, this introduces a new "skipTests" argument that allows users to pass in test names to skip.

### CODE REVIEW

File: pkg/serverstate/statetest/statetest.go [hunk 1]
```diff
@@ -42,6 +45,20 @@ func Test(t *testing.T, f Factory, rf RestartFactory) {
...
+					if name == skipTest {
+						skip = true
+						break
+					}
+				}
+				if skip {
+					t.Run(name, func(t *testing.T) {
+						t.Skip("Test is on the skip list - ignoring")
+					})
+					continue
```
Comment (catsby) [inline pkg/serverstate/statetest/statetest.go:59]: nit/suggestion: we could put a `Loop:` above the `range funcs` on line 42, then simplify this skip check to something like this:

```go
Loop:
  for _, tf := range funcs {
    // existing code
    // ...
    for _, skipTest := range skipTests {
      if name == skipTest {
        t.Run(name, func(t *testing.T) {
          t.Skip("Test is on the skip list - ignoring")
        })
        continue Loop
      }
    }
```

Just a suggestion though, take or leave as you will :)

File: pkg/serverstate/statetest/statetest.go [hunk 2]
```diff
@@ -33,7 +33,10 @@ type (
 // Test runs a validation test suite for a state implementation. All
 // state implementations should pass this suite with no errors to ensure
 // the correct behavior of the state when Waypoint uses it.
-func Test(t *testing.T, f Factory, rf RestartFactory) {
+// skipTests are function names of tests in the serverstate package to skip
+// (i.e. TestJobCreate_singleton). It cannot skip sub-tests (called by
+// t.Run() inside a top-level test)
+func Test(t *testing.T, f Factory, rf RestartFactory, skipTests []string) {
```
Comment (catsby) [inline pkg/serverstate/statetest/statetest.go:39]: How is `skipTests` populated? Could it be populated similar to how the current `tests` slice is?

**Review by catsby — COMMENTED** [review]
Question on where `skipTests` comes from

File: pkg/serverstate/statetest/statetest.go (hunk 1 above)
Comment (catsby) [reply pkg/serverstate/statetest/statetest.go:59]: Ex: https://go.dev/play/p/QYk6_P15hdv

File: pkg/serverstate/statetest/statetest.go (hunk 2 above)
Comment (izaaklauer) [reply pkg/serverstate/statetest/statetest.go:39]: The workflow I envision is that you'd run the tests from a `_test` file in your project, see one that's failing, decide that it's OK from your perspective, and add that test name to the slice where you invoke this `Test` function.

Comment (catsby) [reply pkg/serverstate/statetest/statetest.go:39]: 👍 for posterity, the `skipTests` slice would be added here: 

-https://github.com/hashicorp/waypoint/blob/d4245103aecefae370587efb92695398e21d1bac/internal/server/boltdbstate/state_test.go#L20

I think we need to add a `nil` or empty slice there, yeah?

File: pkg/serverstate/statetest/statetest.go (hunk 1 above)
Comment (izaaklauer) [reply pkg/serverstate/statetest/statetest.go:59]: Ha, I did think about that. I go back and forth on loop labels - they're convenient, but not my favorite for readability.

File: pkg/serverstate/statetest/statetest.go [hunk 3]
```diff
@@ -42,6 +45,20 @@ func Test(t *testing.T, f Factory, rf RestartFactory) {
...
+				skip := false
+				for _, skipTest := range skipTests {
+					if name == skipTest {
+						skip = true
+						break
+					}
+				}
+				if skip {
+					t.Run(name, func(t *testing.T) {
+						t.Skip("Test is on the skip list - ignoring")
```
Comment (briancain) [inline pkg/serverstate/statetest/statetest.go:57]: Can we print the name too? I think that will be helpful if anyone is ever confused why their test didn't run.

**Review by briancain — CHANGES_REQUESTED** [review]

**Review by briancain — APPROVED** [review]
🧪 ✅

**Pull request merged**